liveness: Convert to using hlc.Timestamp

andrewbaptist opened the pull request:
Previously there were some conversions back and forth between hlc times and go times. We always use hlc time when looking at the validity of liveness, and this change makes that more clear.

Epic: none

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/101914)
<!-- Reviewable:end -->

kvoli reviewed (commented):
I've looked through the changes and I'm surprised this updates the splits and to some extent the allocator.

Why is it necessary to thread these through in such places? For example, the split pkg has little dependency on outside pkgs - introducing hlc timestamps seems unnecessary / complex.  What are we gaining here apart from uniformity? The split pkg itself doesn't maintain a clock.

If the HLC timestamps that are being modified are becoming synthetic. does that have implications for accidentally using one for other purposes?

Could @nvanbenschoten take a quick look as well.

> Reviewed 70 of 70 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/101914#-:-NV1riZc7eOFc9SVk4FW:b-oxtym3)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andrewbaptist)

___
*[`pkg/kv/kvserver/replica_consistency.go` line 168 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/101914#-NV1iAmB3gUkRdre47Dc:-NV1iAmB3gUkRdre47Dd:b-amn4zo) ([raw file](https://github.com/cockroachdb/cockroach/blob/8efb1ac9243cb935d292e43cef97b983da876bcf/pkg/kv/kvserver/replica_consistency.go#L168)):*
> ```Go
> 	// stored stats.
> 	//
> 	// Both Persisted and Sub stats were computed deterministically from the
> ```

We use delta below - why change to sub?
___
*[`pkg/kv/kvserver/replica_split_load.go` line 105 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/101914#-NV1qNvt1mrO0IIi7BRT:-NV1qNvt1mrO0IIi7BRU:b-7e7mlz) ([raw file](https://github.com/cockroachdb/cockroach/blob/8efb1ac9243cb935d292e43cef97b983da876bcf/pkg/kv/kvserver/replica_split_load.go#L105)):*
> ```Go
> // find the midpoint based on recorded load.
> func (c *replicaSplitConfig) NewLoadBasedSplitter(
> 	startTime hlc.Timestamp, obj split.SplitObjective,
> ```

Why include a hlc timestamp here as opposed to Time.Time{}?
___
*[`pkg/util/hlc/timestamp.go` line 58 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/101914#-NV1su_8955PRklfzHuT:-NV1su_8955PRklfzHuU:b4rjkhg) ([raw file](https://github.com/cockroachdb/cockroach/blob/8efb1ac9243cb935d292e43cef97b983da876bcf/pkg/util/hlc/timestamp.go#L58)):*
> ```Go
> 
> // After returns whether the provided timestamp is after our timestamp.
> // This matches the
> ```

Comment ends prematurely here.


<!-- Sent from Reviewable.io -->

nvb reviewed (commented):
> Reviewed 70 of 70 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/101914)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andrewbaptist and @kvoli)

___
*[`pkg/kv/kvserver/replica_split_load.go` line 105 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/101914#-NV1qNvt1mrO0IIi7BRT:-NVFTWLvBhsgoO8W3V7N:b-qzbg60) ([raw file](https://github.com/cockroachdb/cockroach/blob/8efb1ac9243cb935d292e43cef97b983da876bcf/pkg/kv/kvserver/replica_split_load.go#L105)):*
<details><summary><i>Previously, kvoli (Austen) wrote…</i></summary><blockquote>

Why include a hlc timestamp here as opposed to Time.Time{}?
</blockquote></details>

Agreed, I don't understand the use of `hlc.Timestamp` in the load-based splitter when all it needs is a wall clock and has no distributed interactions that might benefit from the properties of an HLC. It's worth keeping in mind with this work that HLC readings are more expensive than wall clock readings, so we should use them where needed, but nowhere else.
___
*[`pkg/util/hlc/timestamp.go` line 219 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/101914#-NVFRf1T7pbx69AdaEey:-NVFRf1T7pbx69AdaEez:b-ryb1k9) ([raw file](https://github.com/cockroachdb/cockroach/blob/8efb1ac9243cb935d292e43cef97b983da876bcf/pkg/util/hlc/timestamp.go#L219)):*
> ```Go
> // hlc.Timestamp in the future (or past).
> func (t Timestamp) AddDuration(duration time.Duration) Timestamp {
> 	return Timestamp{
> ```

Can we implement this as `t.Add(dur.Nanoseconds(), logical)`?
___
*[`pkg/util/hlc/timestamp.go` line 222 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/101914#-NVFRnjkAYj5iG6QHtkP:-NVFRnjkAYj5iG6QHtkQ:bpxe0q6) ([raw file](https://github.com/cockroachdb/cockroach/blob/8efb1ac9243cb935d292e43cef97b983da876bcf/pkg/util/hlc/timestamp.go#L222)):*
> ```Go
> 		WallTime:  t.WallTime + duration.Nanoseconds(),
> 		Logical:   t.Logical,
> 		Synthetic: true,
> ```

Synthetic timestamps are going away (https://github.com/cockroachdb/cockroach/issues/101938) now that their role has been replaced entirely by `ClockTimestamp`. If this is causing problems, we can omit it from the patch (you'll get that for free using `Add`). We don't want these synthetic timestamps ending up anywhere new before they are removed.
___
*[`pkg/util/hlc/timestamp.go` line 226 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/101914#-NVFSBu53-qxIT8Yp0QO:-NVFSBu53-qxIT8Yp0QP:b-4ylvdt) ([raw file](https://github.com/cockroachdb/cockroach/blob/8efb1ac9243cb935d292e43cef97b983da876bcf/pkg/util/hlc/timestamp.go#L226)):*
> ```Go
> }
> 
> // Sub returns the delta of this time from a previous timestamp.
> ```

What do we use this method for and what is its desired meaning? The method does not consider the logical portion of the hlc. More importantly, it implies that a clock with the `previous` reading will advance to `t` after the returned `time.Duration`. This is not true, given how HLC clocks advance.


<!-- Sent from Reviewable.io -->

Commit:
liveness: Convert to using hlc.Timestamp

Previously there were some conversions back and forth between hlc times
and go times. We always use hlc time when looking at the validity of
liveness, and this change makes that more clear.

Epic: none

Release note: None

andrewbaptist reviewed (commented):
TFTR. I reverted all changes to split/merge and the asim package. You are correct that it wasn't necessary. In terms of the synthetic timestamp. I also changed to use the add method so it is more clear what is happening. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/101914#-:-NVWFby67mNffh_VGkSs:blc7408)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @kvoli and @nvanbenschoten)

___
*[`pkg/util/hlc/timestamp.go` line 58 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/101914#-NV1su_8955PRklfzHuT:-NVZm5qc8yJd6qXPxBVH:b-dwiyr5) ([raw file](https://github.com/cockroachdb/cockroach/blob/8efb1ac9243cb935d292e43cef97b983da876bcf/pkg/util/hlc/timestamp.go#L58)):*
<details><summary><i>Previously, kvoli (Austen) wrote…</i></summary><blockquote>

Comment ends prematurely here.
</blockquote></details>

Fixed the comment.
___
*[`pkg/util/hlc/timestamp.go` line 219 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/101914#-NVFRf1T7pbx69AdaEey:-NVZmU279-W2LMUkqIGt:bi1lfbf) ([raw file](https://github.com/cockroachdb/cockroach/blob/8efb1ac9243cb935d292e43cef97b983da876bcf/pkg/util/hlc/timestamp.go#L219)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Can we implement this as `t.Add(dur.Nanoseconds(), logical)`?
</blockquote></details>

Yes - changed.
___
*[`pkg/util/hlc/timestamp.go` line 222 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/101914#-NVFRnjkAYj5iG6QHtkP:-NVZmWiY5J2EiLR0cM1F:bdzkyvb) ([raw file](https://github.com/cockroachdb/cockroach/blob/8efb1ac9243cb935d292e43cef97b983da876bcf/pkg/util/hlc/timestamp.go#L222)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Synthetic timestamps are going away (https://github.com/cockroachdb/cockroach/issues/101938) now that their role has been replaced entirely by `ClockTimestamp`. If this is causing problems, we can omit it from the patch (you'll get that for free using `Add`). We don't want these synthetic timestamps ending up anywhere new before they are removed.
</blockquote></details>

No longer relevant as this is using the Add method.
___
*[`pkg/util/hlc/timestamp.go` line 226 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/101914#-NVFSBu53-qxIT8Yp0QO:-NVZmfGq-mxVav22_WxL:b7ao6al) ([raw file](https://github.com/cockroachdb/cockroach/blob/8efb1ac9243cb935d292e43cef97b983da876bcf/pkg/util/hlc/timestamp.go#L226)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

What do we use this method for and what is its desired meaning? The method does not consider the logical portion of the hlc. More importantly, it implies that a clock with the `previous` reading will advance to `t` after the returned `time.Duration`. This is not true, given how HLC clocks advance.
</blockquote></details>

Method removed. It was meant to signify the delta between two timestamps, however this isn't needed anymore and I agree it is somewhat confusing since you shouldn't really be subtracting timestamps. It isn't needed anymore since some of the other changes are gone. 
___
*[`pkg/kv/kvserver/replica_consistency.go` line 168 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/101914#-NV1iAmB3gUkRdre47Dc:-NVZlj9o2CBsEZT18KRU:b-4szj0k) ([raw file](https://github.com/cockroachdb/cockroach/blob/8efb1ac9243cb935d292e43cef97b983da876bcf/pkg/kv/kvserver/replica_consistency.go#L168)):*
<details><summary><i>Previously, kvoli (Austen) wrote…</i></summary><blockquote>

We use delta below - why change to sub?
</blockquote></details>

Sorry, this was my "refactor" getting overly aggressive - reverted. 
___
*[`pkg/kv/kvserver/replica_split_load.go` line 105 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/101914#-NV1qNvt1mrO0IIi7BRT:-NVZn-iT7iOqXhOz_gdB:bp58gk2) ([raw file](https://github.com/cockroachdb/cockroach/blob/8efb1ac9243cb935d292e43cef97b983da876bcf/pkg/kv/kvserver/replica_split_load.go#L105)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Agreed, I don't understand the use of `hlc.Timestamp` in the load-based splitter when all it needs is a wall clock and has no distributed interactions that might benefit from the properties of an HLC. It's worth keeping in mind with this work that HLC readings are more expensive than wall clock readings, so we should use them where needed, but nowhere else.
</blockquote></details>

Reverted the changes to split / merge. 


<!-- Sent from Reviewable.io -->

kvoli reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

Minor question on string formatting in the storepool.

> Reviewed 57 of 57 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/101914#-:-NW3McHGEJvffvSFPU5a:b-hb451w)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @andrewbaptist and @nvanbenschoten)

___
*[`pkg/kv/kvserver/allocator/storepool/store_pool.go` line 545 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/101914#-NW3MU5Q7wI1-XrKvb2V:-NW3MU5Q7wI1-XrKvb2W:b-c9cob) ([raw file](https://github.com/cockroachdb/cockroach/blob/d81dbcc86cb020f6db242d9975dd720c218419d8/pkg/kv/kvserver/allocator/storepool/store_pool.go#L545)):*
> ```Go
> 		}
> 		if detail.ThrottledUntil.After(now) {
> 			fmt.Fprintf(&buf, " [throttled=%v]", detail.ThrottledUntil)
> ```

Will this show the timestamp as nanos now?


<!-- Sent from Reviewable.io -->

andrewbaptist commented:
> Will this show the timestamp as nanos now?

It would have.  I changed it so the formatting doesn't change at all now.

andrewbaptist commented:
bors r=kvoli
